qcore/asserts.py

```diff
@@ -138,10 +138,19 @@ def _dict_path_string(path):
     return "->".join(map(ascii, path))
 
 
+def _to_concrete_dict(value):
+    """Convert dict-subclass instances to concrete dict instances."""
+    assert_is_instance(value, dict)
+    # Not touching already-concrete-dict instances, for performance.
+    if isinstance(value, dict) and type(value) != dict:
+        value = {k: _to_concrete_dict(v) for k, v in value.items()}
```

Comment:
I don't think we need to recurse into the values, because we recurse into assert_dict_eq below already.

Reply:
Oh, right! For this case, we can do only a single-level swap. And it won't have the problem I mentioned earlier, then. So, I guess it's the winner?